keps/sig-node/4781-kublet-restart-pod-status/README.md

```diff
@@ -0,0 +1,705 @@
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+When a kubelet restarts, pods with configured readiness probes may momentarily report containerNotReady status, potentially leading to service disruptions. This KEP proposes enhancing kubelet's pod readiness management to ensure that pod readiness status is preserved across kubelet restarts.
+
+## Motivation
+
+Maintaining high availability and minimizing service interruptions is crucial for Kubernetes clusters. When the kubelet restarts, it re-evaluates the state of all containers and pods. This can sometimes lead to a situation where the container's Ready state, which was previously set based on the success of its readiness probes, gets reset. This reset can cause services to be incorrectly marked as unavailable, even though they were functioning correctly before the kubelet restart.
+
+### Goals
+
+- Maintain the previous ready state of containers after kubelet restart until a new probe result is available.
+- Ensure consistency in container ready states across kubelet restarts.
+- Minimize unnecessary service disruptions caused by temporary ready state changes.
+
+### Non-Goals
```

Comment:
what represents the kubelet restart? How long can it be between stop and start? Do we preserve ready state after VM reboot and how do we differentiate restart and reboot?

Reply:
There's another issue: if the kubelet takes too long to restart, causing the node's status to become `NotReady` (currently the default is 40 seconds), do we still need to ensure that the container's state remains consistent before and after the restart? 
Or should we consider it as a "restart" as long as the node's status hasn't changed?

Reply:
+1 for outlining thses cases

Reply:
- If the container has encountered an issue and startTime is Nil, both the Start and Ready states will be determined based on the probe (as per existing logic)
- If the kubelet's restart time is short, the Container Start and Ready states will both be True(the previous status).
- If the kubelet's restart time is long (exceeding `node-monitor-grace-period`), but the container is healthy, Start will be True and Ready will be False.

Reply:
We compare the startup times of the Kubelet and the container. In the event of a VM reboot, the container will start later, thus the status will be determined by the Startup Probe and Readiness Probe, as the current logic.

Reply:
have you added these considerations to the document?

Reply:
Yeah I mentioned them in the latest version of the KEP.